src/nspawn/nspawn.c

```diff
@@ -2523,6 +2524,23 @@ static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo
         return 0;
 }
 
+static int on_sigchld(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata) {
+        pid_t pid;
+
+        // Because we set PR_SET_CHILD_REAPER, orphan processes may be
+        // reparented to us.
+        pid = PTR_TO_PID(userdata);
+        if (pid == si->ssi_pid || (arg_clone_ns_flags & CLONE_NEWPID)) {
+                // The child has exited. nspawn can terminate now as well.
+                sd_event_exit(sd_event_source_get_event(s), 0);
+        } else {
+                // Orphan. Remove the zombie the from process table.
+                waitpid(si->ssi_pid, NULL, WNOHANG);
+        }
+
+        return 0;
+}
+
```

Comment:
sorry for not noticing this earlier, but this is actually racy because UNIX is stupid... signals like SIGCHLD are not properly queued: they instead simply have "pending" bit that is set or not. This means if SIGCHLD is received twice before our signal handler is called then we'll be dispatched only a single time. This also means the signalfd_siginfo structure will only record the second invocation. Long story short: if two child processes die quickly, one after another, we might only see one SIGCHLD for them, and only reap the second one but leave the first one 
as a zombie...

The only way process this all properly, is using waitid() in a loop, with WNOHANG|WNOWAIT: then decide whether the returned process is one we should reap. if so, call waitid() again, but without WNOHANG and with the PID field filled in. Then, repeat this all until we get ECHILD (and hence know that all our children that have to be reaped have been reaped). 

Hence, something like this:

```c
for (;;) {
        struct sig_info_t si = {};
        if (waitid(P_ALL, 0, &si, WNOHANG|WNOWAIT) < 0)
                return log_error_errno("Failed to wait(): %m");
        if (si.si_pid == 0) /* All pending children are reaped */
                break;
        if (si.si_pid == PTR_TO_PID(userdata)) { /* The main process we care for */
                sd_event_exit(…);
                break;
         }
         /* Other kid we should reap */
         (void) waitid(P_PID, si.si_pid, &si, 0);
}
```